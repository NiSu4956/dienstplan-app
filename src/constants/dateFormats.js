export const DATE_FORMATS = {
  DE_SHORT: 'dd.MM.yyyy',
  DE_LONG: 'EEEE, dd. MMMM yyyy',
  DE_TIME: 'HH:mm',
  ISO: 'yyyy-MM-dd',
  ISO_WITH_TIME: "yyyy-MM-dd'T'HH:mm:ss.SSSxxx",
  WEEK: "'KW' ww '('dd.MM - dd.MM.yyyy')'",
  WEEKDAY_DATE: 'EEEE, dd.MM.yyyy',
  TIME_24: 'HH:mm',
  DAY_MONTH: 'dd.MM',
  MONTH_YEAR: 'MMMM yyyy',
  WEEKDAY: 'EEEE'
};

export const MONTHS = [
  'Januar', 'Februar', 'MÃ¤rz', 'April', 'Mai', 'Juni',
  'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember'
];

export const DAYS_OF_WEEK = [
  'Montag', 'Dienstag', 'Mittwoch', 'Donnerstag', 
  'Freitag', 'Samstag', 'Sonntag'
];

export const TIME_SLOTS = [
  '07:00', '08:00', '09:00', '10:00', '11:00', '12:00', 
  '13:00', '14:00', '15:00', '16:00', '17:00', '18:00', 
  '19:00', '20:00', '21:00'
]; 