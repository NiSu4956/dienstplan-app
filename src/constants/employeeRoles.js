export const EMPLOYEE_ROLES = {
  FULL_TIME: {
    id: 'FULL_TIME',
    name: 'Vollzeit',
    weeklyHours: 40,
    description: 'Vollzeitbeschäftigte Mitarbeiter (40 Stunden/Woche)',
    color: '#4CAF50'
  },
  PART_TIME_30: {
    id: 'PART_TIME_30',
    name: 'Teilzeit 30h',
    weeklyHours: 30,
    description: 'Teilzeitbeschäftigte Mitarbeiter (30 Stunden/Woche)',
    color: '#2196F3'
  },
  PART_TIME_20: {
    id: 'PART_TIME_20',
    name: 'Teilzeit 20h',
    weeklyHours: 20,
    description: 'Teilzeitbeschäftigte Mitarbeiter (20 Stunden/Woche)',
    color: '#9C27B0'
  },
  MINI_JOB: {
    id: 'MINI_JOB',
    name: 'Minijob',
    weeklyHours: 10,
    description: 'Geringfügig beschäftigte Mitarbeiter (max. 10 Stunden/Woche)',
    color: '#FF9800'
  },
  TRAINEE: {
    id: 'TRAINEE',
    name: 'Auszubildende',
    weeklyHours: 40,
    description: 'Auszubildende Mitarbeiter (40 Stunden/Woche)',
    color: '#795548'
  },
  INTERN: {
    id: 'INTERN',
    name: 'Praktikant',
    weeklyHours: 40,
    description: 'Praktikanten (40 Stunden/Woche)',
    color: '#607D8B'
  }
};

export const getEmployeeRole = (roleId) => {
  return EMPLOYEE_ROLES[roleId] || null;
};

export const getWeeklyHours = (roleId) => {
  const role = getEmployeeRole(roleId);
  return role ? role.weeklyHours : 0;
};

export const calculateMonthlyTargetHours = (roleId, year, month) => {
  const weeklyHours = getWeeklyHours(roleId);
  if (!weeklyHours) return 0;

  // Berechne die Anzahl der Arbeitstage im Monat (Mo-Fr)
  const startDate = new Date(year, month, 1);
  const endDate = new Date(year, month + 1, 0);
  let workingDays = 0;

  for (let d = startDate; d <= endDate; d.setDate(d.getDate() + 1)) {
    // 0 = Sonntag, 6 = Samstag
    if (d.getDay() !== 0 && d.getDay() !== 6) {
      workingDays++;
    }
  }

  // Berechne die Zielstunden für den Monat
  // (Wochenstunden / 5 Arbeitstage) * Anzahl der Arbeitstage im Monat
  return (weeklyHours / 5) * workingDays;
}; 