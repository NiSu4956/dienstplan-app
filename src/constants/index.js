export const DAYS_OF_WEEK = ['Montag', 'Dienstag', 'Mittwoch', 'Donnerstag', 'Freitag', 'Samstag', 'Sonntag'];

export const TIME_SLOTS = ['7:00', '8:00', '9:00', '10:00', '11:00', '12:00', '13:00', '14:00', 
                     '15:00', '16:00', '17:00', '18:00', '19:00', '20:00', '21:00'];

export const STORAGE_KEYS = {
  USER: 'currentUser',
  SCHEDULE_DATA: 'scheduleData',
  CHILDREN: 'children'
};

export const ROLES = {
  ADMIN: 'admin'
};

export const REQUEST_TYPES = {
  VACATION: 'vacation',
  SICK: 'sick'
};

export const DEFAULT_TIME_SLOT = '07:00';

export const REQUEST_STATUS = {
  PENDING: 'pending',
  APPROVED: 'approved',
  REJECTED: 'rejected'
};

export const REQUEST_STATUS_TEXT = {
  pending: 'Ausstehend',
  approved: 'Genehmigt',
  rejected: 'Abgelehnt'
};

export const REQUEST_STATUS_CLASS = {
  pending: 'status-pending',
  approved: 'status-approved',
  rejected: 'status-rejected'
};

export const PDF_STYLES = {
  FONT_SIZES: {
    TITLE: 16,
    NORMAL: 8,
    FOOTER: 8
  },
  COLORS: {
    BLUE: [79, 70, 229],
    WHITE: 255,
    GRAY: [245, 245, 245],
    SHIFT_COLORS: {
      EARLY: [219, 234, 254],
      DAY: [220, 252, 231],
      LATE: [243, 232, 255],
      NIGHT: [229, 231, 235],
      COOKING: [254, 226, 226],
      WEEKEND: [254, 249, 195]
    }
  },
  MARGINS: {
    TOP: 20,
    RIGHT: 14,
    BOTTOM: 20,
    LEFT: 14
  },
  CELL_STYLES: {
    PADDING: 2,
    FIRST_COLUMN_WIDTH: {
      DAY_VIEW: 20,
      WEEK_VIEW: 15
    }
  }
}; 