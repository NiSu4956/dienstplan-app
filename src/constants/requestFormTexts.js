export const REQUEST_TYPES = {
  VACATION: 'vacation',
  SICK: 'sickness'
};

export const CUSTOM_SHIFTS = {
  VACATION: {
    title: 'üèñÔ∏è Urlaub',
    color: 'blue',
    startTime: '00:00',
    endTime: '23:59'
  },
  SICK: {
    title: 'üè• Krank',
    color: 'red',
    startTime: '00:00',
    endTime: '23:59'
  }
};

export const REQUEST_FORM_TEXTS = {
  labels: {
    startDate: 'Startdatum:',
    endDate: 'Enddatum:',
    notes: 'Anmerkungen:'
  },
  placeholders: {
    vacation: 'Zus√§tzliche Informationen zum Urlaub...',
    sickness: 'Grund der Krankmeldung...'
  },
  buttons: {
    cancel: 'Abbrechen',
    submit: {
      vacation: 'Urlaub beantragen',
      sickness: 'Krankmeldung einreichen'
    }
  },
  error: {
    title: 'Fehler',
    validation: {
      invalidDateRange: 'Das Enddatum muss nach dem Startdatum liegen.',
      invalidDate: 'Bitte w√§hlen Sie ein g√ºltiges Datum.',
      invalidShift: 'Der gew√§hlte Zeitraum √ºberschneidet sich mit geplanten Schichten.',
      invalidRequest: 'Ung√ºltige Anfrage. Bitte √ºberpr√ºfen Sie Ihre Eingaben.',
      pastDate: 'Das Startdatum darf nicht in der Vergangenheit liegen.',
      maxDuration: 'Die maximale Dauer f√ºr einen Antrag betr√§gt {days} Tage.',
      requiredFields: 'Bitte f√ºllen Sie alle erforderlichen Felder aus.',
      shiftConflict: 'Sie sind am {date} bereits f√ºr eine Schicht eingeplant. Ein Urlaubsantrag ist f√ºr diesen Tag nicht m√∂glich.',
      existingVacation: 'Sie haben bereits einen genehmigten Urlaubsantrag f√ºr den {date}. Bitte w√§hlen Sie einen anderen Zeitraum.'
    }
  }
};

export const DATE_PICKER_CONFIG = {
  dateFormat: 'dd.MM.yyyy',
  locale: 'de',
  className: 'form-control'
}; 