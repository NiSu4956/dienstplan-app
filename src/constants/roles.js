/**
 * Rollen-Konstanten f√ºr die Anwendung
 */
export const ROLES = {
  ADMIN: 'admin',
  EMPLOYEE: 'employee'
}; 