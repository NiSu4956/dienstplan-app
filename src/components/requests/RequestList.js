import React from 'react';
import { formatDate } from '../../utils/dateUtils';
import { REQUEST_TYPES, REQUEST_STATUS_TEXT, REQUEST_STATUS_CLASS } from '../../constants';

function RequestList({ requests, title }) {
  if (!requests || requests.length === 0) {
    return null;
  }

  return (
    <div className="request-list">
      <h4>{title}</h4>
      {requests.map(request => (
        <div key={request.id} className="request-card">
          <div className="request-header">
            <span className={`request-type ${request.type}`}>
              {request.type === REQUEST_TYPES.VACATION ? 'Urlaub' : 'Krankmeldung'}
            </span>
            <span className={`request-status ${REQUEST_STATUS_CLASS[request.status]}`}>
              {REQUEST_STATUS_TEXT[request.status]}
            </span>
          </div>
          <div className="request-dates">
            {formatDate(request.startDate)} - {formatDate(request.endDate)}
          </div>
          {request.notes && (
            <div className="request-notes">
              {request.notes}
            </div>
          )}
          {request.adminComment && (
            <div className="admin-comment">
              <strong>Kommentar:</strong> {request.adminComment}
            </div>
          )}
          <div className="request-timestamp">
            Eingereicht am: {formatDate(request.submittedAt)}
          </div>
        </div>
      ))}
    </div>
  );
}

export default RequestList; 