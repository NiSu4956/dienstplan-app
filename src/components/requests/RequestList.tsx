import React from 'react';
import { formatDate } from '../../utils/dateUtils';
import { REQUEST_TYPES, REQUEST_STATUS_TEXT, REQUEST_STATUS_CLASS } from '../../constants';

type RequestStatus = keyof typeof REQUEST_STATUS_TEXT;

interface Request {
  id?: string;
  type: typeof REQUEST_TYPES[keyof typeof REQUEST_TYPES];
  status: RequestStatus;
  employeeName?: string;
  employeeId?: string;
  startDate: string;
  endDate: string;
  notes?: string;
  adminComment?: string;
  submittedAt?: string;
}

interface RequestListProps {
  requests: Request[];
  title: string;
}

function RequestList({ requests, title }: RequestListProps) {
  if (!requests || requests.length === 0) {
    return null;
  }

  return (
    <div className="request-list">
      <h4>{title}</h4>
      {requests.map(request => (
        <div key={request.id} className="request-card">
          <div className="request-header">
            <span className={`request-type ${request.type}`}>
              {request.type === REQUEST_TYPES.VACATION ? 'Urlaub' : 'Krankmeldung'}
            </span>
            <span className={`request-status ${REQUEST_STATUS_CLASS[request.status]}`}>
              {REQUEST_STATUS_TEXT[request.status]}
            </span>
          </div>
          <div className="request-dates">
            {formatDate(request.startDate)} - {formatDate(request.endDate)}
          </div>
          {request.notes && (
            <div className="request-notes">
              {request.notes}
            </div>
          )}
          {request.adminComment && (
            <div className="admin-comment">
              <strong>Kommentar:</strong> {request.adminComment}
            </div>
          )}
          <div className="request-timestamp">
            Eingereicht am: {formatDate(request.submittedAt || '')}
          </div>
        </div>
      ))}
    </div>
  );
}

export default RequestList; 