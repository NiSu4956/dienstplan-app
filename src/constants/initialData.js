import { ROLES } from './roles';

export const INITIAL_SCHEDULE_DATA = {};

export const INITIAL_EMPLOYEES = [
  { id: 1, name: 'Sabine', role: 'Vollzeit', qualifications: ['WG1', 'WG2', 'Nachtdienst'], workingHours: 40 },
  { id: 2, name: 'Manu', role: 'Vollzeit', qualifications: ['WG1', 'Kochen'], workingHours: 40 },
  { id: 3, name: 'Levin', role: 'Teilzeit', qualifications: ['Schule', 'Freizeitaktivitäten'], workingHours: 20 },
  { id: 7, name: 'Eva', role: 'Teilzeit', qualifications: ['WG1', 'Nachmittagsprogramm'], workingHours: 25 },
  { id: 8, name: 'Fabi', role: 'Vollzeit', qualifications: ['WG2', 'Nachtdienst'], workingHours: 40 },
  { id: 10, name: 'Admin', role: ROLES.ADMIN, qualifications: ['Administration'], workingHours: 40, password: 'Admin' }
];

export const INITIAL_CHILDREN = [
  { 
    id: 1, 
    name: 'Max Mustermann', 
    group: 'WG1',
    birthDate: '2018-05-15',
    notes: 'Allergisch gegen Erdnüsse',
    documentation: []
  },
  { 
    id: 2, 
    name: 'Lisa Schmidt', 
    group: 'WG2',
    birthDate: '2019-03-22',
    notes: 'Nimmt regelmäßig Medikamente',
    documentation: []
  }
];

export const INITIAL_SHIFT_TYPES = [
  { id: 1, name: 'Frühdienst', startTime: '07:00', endTime: '14:00', color: 'blue' },
  { id: 2, name: 'Tagesdienst', startTime: '09:00', endTime: '17:00', color: 'green' },
  { id: 3, name: 'Spätdienst', startTime: '14:00', endTime: '21:00', color: 'purple' },
  { id: 4, name: 'Nachtdienst', startTime: '21:00', endTime: '07:00', color: 'gray' },
  { id: 5, name: 'Kochen', startTime: '11:00', endTime: '14:00', color: 'red' },
  { id: 6, name: 'Wochenende', startTime: '09:00', endTime: '21:00', color: 'yellow' }
]; 