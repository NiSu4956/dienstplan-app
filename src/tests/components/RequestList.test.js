import React from 'react';
import { render } from '@testing-library/react';
import RequestList from '../../components/requests/RequestList';
import { REQUEST_TYPES } from '../../constants';

describe('RequestList', () => {
  const mockRequests = [
    {
      id: 1,
      type: REQUEST_TYPES.VACATION,
      status: 'approved',
      startDate: '2024-03-15',
      endDate: '2024-03-20',
      notes: 'Urlaub in den Bergen',
      submittedAt: '2024-03-01T10:00:00',
      adminComment: 'Genehmigt'
    },
    {
      id: 2,
      type: REQUEST_TYPES.SICK,
      status: 'pending',
      startDate: '2024-03-16',
      endDate: '2024-03-17',
      notes: 'Grippe',
      submittedAt: '2024-03-16T08:00:00'
    }
  ];

  it('sollte null zurückgeben, wenn keine Requests vorhanden sind', () => {
    const { container } = render(<RequestList requests={[]} title="Test" />);
    expect(container.firstChild).toBeNull();
  });

  it('sollte die Liste der Requests korrekt rendern', () => {
    const { container } = render(
      <RequestList 
        requests={mockRequests} 
        title="Urlaubsanträge" 
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('sollte die korrekten Status-Badges anzeigen', () => {
    const { getByText } = render(
      <RequestList 
        requests={mockRequests} 
        title="Urlaubsanträge" 
      />
    );
    
    const statusBadge = getByText('Genehmigt', { selector: '.request-status' });
    expect(statusBadge).toBeInTheDocument();
    expect(getByText('Ausstehend', { selector: '.request-status' })).toBeInTheDocument();
  });

  it('sollte die korrekten Request-Typen anzeigen', () => {
    const { getByText } = render(
      <RequestList 
        requests={mockRequests} 
        title="Urlaubsanträge" 
      />
    );
    
    const vacationType = getByText('Urlaub', { selector: '.request-type' });
    const sickType = getByText('Krankmeldung', { selector: '.request-type' });
    
    expect(vacationType).toBeInTheDocument();
    expect(sickType).toBeInTheDocument();
  });

  it('sollte Notizen und Admin-Kommentare anzeigen', () => {
    const { getByText } = render(
      <RequestList 
        requests={mockRequests} 
        title="Urlaubsanträge" 
      />
    );
    
    expect(getByText('Urlaub in den Bergen', { selector: '.request-notes' })).toBeInTheDocument();
    expect(getByText('Grippe', { selector: '.request-notes' })).toBeInTheDocument();
    expect(getByText('Kommentar:', { selector: '.admin-comment strong' })).toBeInTheDocument();
    expect(getByText('Genehmigt', { selector: '.admin-comment' })).toBeInTheDocument();
  });
}); 