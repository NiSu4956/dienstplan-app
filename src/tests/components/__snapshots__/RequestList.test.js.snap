// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestList sollte die Liste der Requests korrekt rendern 1`] = `
<div>
  <div
    class="request-list"
  >
    <h4>
      Urlaubsantr√§ge
    </h4>
    <div
      class="request-card"
    >
      <div
        class="request-header"
      >
        <span
          class="request-type vacation"
        >
          Urlaub
        </span>
        <span
          class="request-status status-approved"
        >
          Genehmigt
        </span>
      </div>
      <div
        class="request-dates"
      >
        15.03.2024
         - 
        20.03.2024
      </div>
      <div
        class="request-notes"
      >
        Urlaub in den Bergen
      </div>
      <div
        class="admin-comment"
      >
        <strong>
          Kommentar:
        </strong>
         
        Genehmigt
      </div>
      <div
        class="request-timestamp"
      >
        Eingereicht am: 
        01.03.2024
      </div>
    </div>
    <div
      class="request-card"
    >
      <div
        class="request-header"
      >
        <span
          class="request-type sick"
        >
          Krankmeldung
        </span>
        <span
          class="request-status status-pending"
        >
          Ausstehend
        </span>
      </div>
      <div
        class="request-dates"
      >
        16.03.2024
         - 
        17.03.2024
      </div>
      <div
        class="request-notes"
      >
        Grippe
      </div>
      <div
        class="request-timestamp"
      >
        Eingereicht am: 
        16.03.2024
      </div>
    </div>
  </div>
</div>
`;
